From 428846b2c6381d777c42f83d2b991cab1ad86926 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 27 May 2025 16:05:15 +0300
Subject: [PATCH] [JSW-704] jgs: address space update

With this patch, updated firmware memory filter
api as per new address space.
Removed support to filter 32bit address because
it may wrongly filterout input address when
address space changes and to make it more
accurate.
Updated arc address conversion macro, since
fw mem sizes changed from fs1 and existing
macro not working for new address space.

Change-Id: I775fd007cf222f2ce6932a6e461fe14d873c7854
---

diff --git a/arch/jgs/arc/common.c b/arch/jgs/arc/common.c
index 1ef91ac..0171e51 100644
--- a/arch/jgs/arc/common.c
+++ b/arch/jgs/arc/common.c
@@ -361,36 +361,27 @@
 	return true;
 }
 
-/* This function is only applicable for io die master */
+/*
+ * This API to check particular address in FW Mem All ARC ICCM mem range.
+ * Can be used only for 64Bit address.
+ */
 bool boot_rom_is_valid_address_range(uint64_t addr, uint32_t len)
 {
-	if (addr >= (uint64_t)IO_0_SP_SRAM_BASE &&
-	    addr + len <= (uint64_t)(IO_0_SP_SRAM_BASE + IO_0_SP_SRAM_SIZE))
-		return true;
+	if (CHECK_ADDR_IN_RANGE(addr, len, IO_0_FW_MEM_BASE_FULL, IO_0_FW_MEM_SIZE)) {
 
-	if (addr >= IO_0_SP_SRAM_BASE_FULL &&
-	    addr + len <= IO_0_SP_SRAM_BASE_FULL + IO_0_SP_SRAM_SIZE)
-		return true;
+		if ((CHECK_ADDR_IN_RANGE(addr, len, IO_0_PCIE_ARC_ICCM_BASE_FULL, IO_0_PCIE_ARC_ICCM_SIZE)) ||
+			(CHECK_ADDR_IN_RANGE(addr, len, IO_0_RST_ARC_ICCM_BASE_FULL, IO_0_RST_ARC_ICCM_SIZE)) ||
+			(CHECK_ADDR_IN_RANGE(addr, len, IO_0_MNG_ARC_ICCM_BASE_FULL, IO_0_MNG_ARC_ICCM_SIZE)) ||
+			(CHECK_ADDR_IN_RANGE(addr, len, IO_0_PID_ARC_ICCM_BASE_FULL, IO_0_PID_ARC_ICCM_SIZE)) ||
+			(CHECK_ADDR_IN_RANGE(addr, len, IO_0_MISC_ARC_ICCM_BASE_FULL, IO_0_MISC_ARC_ICCM_SIZE)) ||
+			(CHECK_ADDR_IN_RANGE(addr, len, IO_0_PSOC_FW_SRAM_BASE_FULL, IO_0_FWSRAM_SIZE)) ||
+			(CHECK_ADDR_IN_RANGE(addr, len, IO_0_SP_SRAM_BASE_FULL, IO_0_SP_SRAM_SIZE))) {
 
-	if (addr >= IO_0_RST_ARC_ICCM_BASE_FULL &&
-	    addr + len <= IO_0_RST_ARC_ICCM_BASE_FULL + IO_0_RST_ARC_MEM_SIZE)
-		return true;
+			return true;
+		}
 
-	if (addr >= IO_0_MNG_ARC_ICCM_BASE_FULL &&
-	    addr + len <= IO_0_MNG_ARC_ICCM_BASE_FULL + IO_0_MNG_ARC_MEM_SIZE)
-		return true;
-
-	if (addr >= IO_0_PID_ARC_ICCM_BASE_FULL &&
-	    addr + len <= IO_0_PID_ARC_ICCM_BASE_FULL + IO_0_PID_ARC_MEM_SIZE)
-		return true;
-
-	if (addr >= (uint64_t)(IO_0_PSOC_FW_SRAM_BASE) &&
-	    addr + len <= (uint64_t)(IO_0_PSOC_FW_SRAM_BASE + IO_0_FWSRAM_SIZE))
-		return true;
-
-	if (addr >= IO_0_PSOC_FW_SRAM_BASE_FULL &&
-	    addr + len <= IO_0_PSOC_FW_SRAM_BASE_FULL + IO_0_FWSRAM_SIZE)
-		return true;
+		return false;
+	}
 
 	return false;
 }
@@ -398,9 +389,9 @@
 uint32_t boot_rom_fw_mem_arc_addr(uint64_t addr)
 {
 	if (addr >= IO_0_MNG_ARC_ICCM_BASE_FULL && addr <= IO_0_MNG_ARC_ICCM_BASE_FULL + IO_0_MNG_ARC_MEM_SIZE)
-		return CONV_64B_TO_ARC_FW_MEM(addr) - (IO_0_MNG_ARC_ICCM_BASE);
+		return GET_ARC_ADDR(addr, REGION_FW_MEM_IDX) - (IO_0_MNG_ARC_ICCM_BASE);
 	else
-		return CONV_64B_TO_ARC_FW_MEM(addr);
+		return GET_ARC_ADDR(addr, REGION_FW_MEM_IDX);
 }
 
 uint32_t boot_rom_get_arc_addr(uint64_t addr)
